File: python/04-integration/03-simpson38rule/simpson38rule.py
# Python3 code to implement 
# Simpson's rule 

# Given function to be 
# integrated 
def func(x): 
	
	return (float(1) / ( 1 + x * x )) 


# Function to perform calculations 
def calculate(lower_limit, upper_limit, interval_limit ): 
	
	interval_size = (float(upper_limit - lower_limit) / interval_limit) 
	sum = func(lower_limit) + func(upper_limit); 

	# Calculates value till integral limit 
	for i in range(1, interval_limit ): 
		if (i % 3 == 0): 
			sum = sum + 2 * func(lower_limit + i * interval_size) 
		else: 
			sum = sum + 3 * func(lower_limit + i * interval_size) 
	
	return ((float( 3 * interval_size) / 8 ) * sum ) 

# driver function 
interval_limit = 10
lower_limit = 1
upper_limit = 10

integral_res = calculate(lower_limit, upper_limit, interval_limit) 

# rounding the final answer to 6 decimal places 
print (round(integral_res, 6)) 

# This code is contributed by Saloni. 

File: cpp/04-integration/03-simpson38rule/simpson38rule.cpp
// CPP program to implement Simpson's rule 
#include<iostream> 
using namespace std; 

// Given function to be integrated 
float func( float x) 
{ 
	return (1 / ( 1 + x * x )); 
} 

// Function to perform calculations 
float calculate(float lower_limit, float upper_limit, 
				int interval_limit ) 
{ 
	float value; 
	float interval_size = (upper_limit - lower_limit) 
						/ interval_limit; 
	float sum = func(lower_limit) + func(upper_limit); 

	// Calculates value till integral limit 
	for (int i = 1 ; i < interval_limit ; i++) 
	{ 
		if (i % 3 == 0) 
			sum = sum + 2 * func(lower_limit + i * interval_size); 
		else
			sum = sum + 3 * func(lower_limit + i * interval_size); 
	} 
	return ( 3 * interval_size / 8 ) * sum ; 
} 

// Driver Code 
int main() 
{ 
	int interval_limit = 10; 
	float lower_limit = 1; 
	float upper_limit = 10; 
	float integral_res = calculate(lower_limit, upper_limit, 
								interval_limit); 

	cout << integral_res; 
	return 0; 
} 

